sncosmo/models.py

```diff
@@ -997,6 +998,317 @@ def colorlaw(self, wave=None):
+
+        The second component is block diagonal. It has
+        constant covariance between all photometry points within a
+        bandpass (regardless of phase), and no covariance between
+        photometry points in different bandpasses:
+
+        .. math::
+
+           CD(\\lambda_c)^2
+
+        where the 1-d function :math:`CD` is given by the file ``cdfile``.
+        Adding these two components gives the *relative* covariance on model
+        photometry.
+
+        Parameters
+        ----------
+        band : `~numpy.ndarray` of `~sncosmo.Bandpass`
+            Bandpasses of observations.
+        phase : `~numpy.ndarray` (float)
+            Phases of observations.
+
+
+        Returns
+        -------
+        rcov : `~numpy.ndarray`
+            Model relative covariance for given bandpasses and phases.
+        """
+
+        # construct covariance array with relative variance on diagonal
+        diagonal = np.zeros(phase.shape, dtype=np.float64)
+        for b in set(band):
+            mask = band == b
+            diagonal[mask] = self._bandflux_rvar_single(b, phase[mask])
+        result = np.diagflat(diagonal)
+
+        # add kcorr errors
+        for b in set(band):
+            mask1d = band == b
+            mask2d = mask1d * mask1d[:, None]  # mask for result array
+            kcorrerr = self._colordisp(b.wave_eff)
+            result[mask2d] += kcorrerr**2
+
+        return result
+
+    def _set_colorlaw_from_file(self, name_or_obj):
+        """Read color law file and set the internal colorlaw function."""
+
+        # Read file
+        if isinstance(name_or_obj, str):
+            f = open(name_or_obj, 'r')
+        else:
+            f = name_or_obj
+        words = f.read().split()
+        f.close()
+
+        # Get colorlaw coeffecients.
+        ncoeffs = int(words[0])
+        colorlaw_coeffs = [float(word) for word in words[1: 1 + ncoeffs]]
+
+        # If there are more than 1+ncoeffs words in the file, we expect them to
+        # be of the form `keyword value`.
+        version = 0
+        colorlaw_range = [2800., 8000.]
+        for i in range(1+ncoeffs, len(words), 2):
+            if words[i] == 'Salt2ExtinctionLaw.version':
+                version = int(words[i+1])
+            elif words[i] == 'Salt2ExtinctionLaw.min_lambda':
+                colorlaw_range[0] = float(words[i+1])
+            elif words[i] == 'Salt2ExtinctionLaw.max_lambda':
+                colorlaw_range[1] = float(words[i+1])
+            else:
+                raise RuntimeError("Unexpected keyword: {}".format(words[i]))
+
+        # Set extinction function to use.
+        if version == 0:
+            raise Exception("Salt2ExtinctionLaw.version 0 not supported.")
+        elif version == 1:
+            self._colorlaw = SALT2ColorLaw(colorlaw_range, colorlaw_coeffs)
+        else:
+            raise Exception('unrecognized Salt2ExtinctionLaw.version: ' +
+                            version)
```

Comment:
Minor comment: `RuntimeError` might be the most appropriate exception here. Also, `version = 0` might indicate that the version keyword wasn't found in the file.

Reply:
I removed this code from SALT3Source (see comment above), but changed this in the SALT2Source just for fun

Reply:
That was probably me who wrote `Exception` initially! :smile: